subsys/zigbee/osif/zb_nrf_serial.c

```diff
@@ -391,8 +355,82 @@ void zb_osif_serial_set_cb_send_data(serial_send_data_cb_t cb)
 	tx_data_cb = cb;
 }
 
-void zb_osif_serial_flush(void)
+void zb_osif_async_serial_flush(void)
 {
 	(void)k_sem_take(&tx_done_sem, K_FOREVER);
 	k_sem_give(&tx_done_sem);
 }
+
+#ifdef CONFIG_ZBOSS_TRACE_BINARY_NCP_TRANSPORT_LOGGING
+void zb_osif_set_uart_byte_received_cb(zb_callback_t hnd)
+{
+	char_handler = hnd;
+}
+
+void zb_osif_set_user_io_buffer(zb_byte_array_t *buf_ptr, zb_ushort_t capacity)
+{
+	(void)k_sem_take(&tx_done_sem, K_FOREVER);
+
+	uart_tx_buf = buf_ptr->ring_buf;
+	uart_tx_buf_size = capacity;
+
+	k_sem_give(&tx_done_sem);
+}
+
+void zb_osif_serial_put_bytes(const zb_uint8_t *buf, zb_short_t len)
+{
+#if !(defined(ZB_HAVE_ASYNC_SERIAL) && \
+	defined(CONFIG_ZBOSS_TRACE_LOG_LEVEL_OFF))
+
+	if ((uart_dev == NULL) || is_sleeping) {
+		return;
+	}
+
+	if (len > uart_tx_buf_size) {
+		return;
+	}
+
+	/*
+	 * Wait forever since there is no way to inform higher layer
+	 * about TX busy state.
+	 */
+	(void)k_sem_take(&tx_done_sem, K_FOREVER);
+	memcpy(uart_tx_buf, buf, len);
+
+	uart_tx_buf_len = len;
+	uart_tx_buf_offset = 0;
+
+	/* Enable tx interrupts. */
+	uart_irq_tx_enable(uart_dev);
+
+#endif /* !(ZB_HAVE_ASYNC_SERIAL && CONFIG_ZBOSS_TRACE_LOG_LEVEL_OFF) */
+}
+#endif /* CONFIG_ZBOSS_TRACE_BINARY_NCP_TRANSPORT_LOGGING */
+
+void zb_osif_serial_init(void)
```

Comment:
Personally, I'd prefer to rename this file to `zb_nrf_async_serial.c` and create a new `zb_nrf_serial.c`, included in all variants, that will dispatch calls to the `_init()`, `_sleep()`, `_wake_up` and `_flush()` methods based on Kconfig symbols.
By doing this, it should be possible to remove the stinky `#ifndef CONFIG_ZIGBEE_HAVE_SERIAL` from non-NCP backends.